### PR TITLE
Retry ILM steps that fail due to SnapshotInProgressException

### LINKED ISSUE #37541: ILM Actions can fail due to in-progress snapshots of indices
**situation**

Actions like: Delete, Freeze, Shrink all execute ES actions like
DeleteIndex and CloseIndex that throw IllegalArgumentExceptions in the cases
that snapshots of their respective indices are in progress.

**What can ES do about it?**

A SnapshotInProgressException can be thrown instead of an IAE so 
that ILM can be signaled more explicitly that this is the case so that 
it can retry actions appropriately. Actions should not move to an error 
step.

### PR DESCRIPTION
**Pull request by dakrone**
Some steps, such as steps that delete, close, or freeze an index, may fail due to a currently running snapshot of the index. In those cases, rather than move to the ERROR step, we should retry the step when the snapshot has completed.

This change adds an abstract step (`AsyncRetryDuringSnapshotActionStep`) that certain steps (like the ones I mentioned above) can extend that will automatically handle a situation where a snapshot is taking place. When a `SnapshotInProgressException` is received by the listener wrapper, a `ClusterStateObserver` listener is registered to wait until the snapshot has completed, re-running the ILM action when no snapshot is occurring.

This also adds integration tests for these scenarios (thanks to @talevy in #37552).

Resolves #37541

### CODE REVIEW

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/indexlifecycle/CloseFollowerIndexStepTests.java [hunk 1]
```diff
@@ -114,4 +109,30 @@ public void onFailure(Exception e) {
         Mockito.verify(indicesClient).close(Mockito.any(), Mockito.any());
         Mockito.verifyNoMoreInteractions(indicesClient);
     }
+
+    @Override
+    protected CloseFollowerIndexStep createRandomInstance() {
```
Comment (talevy) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/indexlifecycle/CloseFollowerIndexStepTests.java]: 👍

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/AsyncRetryDuringSnapshotActionStep.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                            IndexMetaData idxMeta = state.metaData().index(index);
+                            if (idxMeta == null) {
+                                // The index has since been deleted, mission accomplished!
+                                originalListener.onResponse(true);
+                            }
+                            // Re-invoke the performAction method with the new state
+                            performAction(idxMeta, state, observer, originalListener);
+                        }, originalListener::onFailure),
+                        // TODO: what is a good timeout value for no new state received during this time?
+                        TimeValue.timeValueHours(12));
```
Comment (talevy) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/AsyncRetryDuringSnapshotActionStep.java]: I think waiting 12 hours for a snapshot to finish is reasonable. If there is no progress on this action in that time interval, a user may want to know. so 👍

**Review by talevy — APPROVED** [review]
LGTM

I discussed this outside of Github with @dakrone, but we agreed
that unit tests for `AsyncRetryDuringSnapshotActionStep`'s 
`SnapshotExceptionListener` and `NoSnapshotRunningListener`
would only cover some non-integral branches of the code logic in 
retrying actions via the cluster-state-observable. Since there
is confidence that the existing integration tests in this PR cover
the successful retry, that represents the critical path and sufficient
for verifying these changes do what they intend.

**Comment by talevy** [comment]
backport to 6.x is blocked on https://github.com/elastic/elasticsearch/pull/37723 (SnapshotInProgressException)

**Comment by talevy** [comment]
update: above blocker PR for 6.x was merged

**Review by AthenaEryma — APPROVED** [review]
LGTM

**Pull request merged**